### PR TITLE
Cas Initializr UI

### PR DESCRIPTION
**Pull request by rmathis**
This adds a user interface based on React to the Initializr UI.

### CODE REVIEW

**Comment by rmathis** [comment]
Will resolve these issues by removing the library in question

**Comment by mmoayyed** [comment]
@rmathis This is most excellent. Well done. 

Follow ups:

1. Please hide the Optional panel, and make it expandable in case the user wants to modify the settings
2. Please use defaults for all optional values

Group: org.apereo.cas
artifact: cas-overlay

What is `Name` here? 

3. Remove the word "demo" from tooltips and other helper hints
4. Package should likely always be WAR, and must not be editable. Initializr only generates one type of packaging.
5. The hamburger menu on the left should expand, should likely provide options/links to the Apereo CAS website, CAS on Github, and CAS documentation website
6. We should likely add some kind of footer to the main page to provide version info on the initializr itself, the typical "powered by". 
7. The type dropdown should just shorten the options to be "CAS Overlay" etc. Remove the words "Gradle Project Zip" from options. 
8. Please provide an option to remove all selected dependencies at once and clear all.
9. The "Dependencies" label seems unnecessary. 
10. Replace "Customization" with "Build your CAS deployment".

At some point in the future, it would be excellent to find a way to browse the contents of the would-be-generated overlay prior to downloading it.

File: ui/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
+
+## Available Scripts
+
+In the project directory, you can run:
+
+### `yarn start`
+
+Runs the app in the development mode.\
```
Comment (mmoayyed) [inline ui/README.md:11]: The backslashes should be removed, no?

**Comment by mmoayyed** [comment]
Also it might be good to think about how available dependencies can be grouped by type and filtered, as you have them labeled.

**Comment by rmathis** [comment]
> What is `Name` here?

It ends up being the name of the file downloaded, so it defaults to `cas.[tar.gz]`. We can remove the option if you'd like, but it was in the api so I included it.

**Comment by rmathis** [comment]
> Also it might be good to think about how available dependencies can be grouped by type and filtered, as you have them labeled.

This has been added

**Comment by rmathis** [comment]
> @rmathis This is most excellent. Well done.
> 
> Follow ups:
> 
>     1. Please hide the Optional panel, and make it expandable in case the user wants to modify the settings
> 
>     2. Please use defaults for all optional values
> 
> 
> Group: org.apereo.cas artifact: cas-overlay
> 
> What is `Name` here?
> 
>     3. Remove the word "demo" from tooltips and other helper hints
> 
>     4. Package should likely always be WAR, and must not be editable. Initializr only generates one type of packaging.
> 
>     5. The hamburger menu on the left should expand, should likely provide options/links to the Apereo CAS website, CAS on Github, and CAS documentation website
> 
>     6. We should likely add some kind of footer to the main page to provide version info on the initializr itself, the typical "powered by".
> 
>     7. The type dropdown should just shorten the options to be "CAS Overlay" etc. Remove the words "Gradle Project Zip" from options.
> 
>     8. Please provide an option to remove all selected dependencies at once and clear all.
> 
>     9. The "Dependencies" label seems unnecessary.
> 
>     10. Replace "Customization" with "Build your CAS deployment".
> 
> 
> At some point in the future, it would be excellent to find a way to browse the contents of the would-be-generated overlay prior to downloading it.

These items have been taken care of with the latest push

**Comment by mmoayyed** [comment]
Latest commit causes the following:

```
2022-11-23 09:34:34.754 DEBUG 28299 --- [nio-8080-exec-8] o.s.web.servlet.DispatcherServlet        : Failed to complete request: org.thymeleaf.exceptions.TemplateInputException: Error resolving template [index], template might not exist or might not be accessible by any of the configured Template Resolvers
2022-11-23 09:34:34.756 ERROR 28299 --- [nio-8080-exec-8] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.thymeleaf.exceptions.TemplateInputException: Error resolving template [index], template might not exist or might not be accessible by any of the configured Template Resolvers] with root cause

org.thymeleaf.exceptions.TemplateInputException: Error resolving template [index], template might not exist or might not be accessible by any of the configured Template Resolvers
        at org.thymeleaf.engine.TemplateManager.resolveTemplate(TemplateManager.java:869) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
        at org.thymeleaf.engine.TemplateManager.parseAndProcess(TemplateManager.java:607) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
        at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1098) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
        at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1072) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
        at org.thymeleaf.spring5.view.ThymeleafView.renderFragment(ThymeleafView.java:366) ~[thymeleaf-spring5-3.0.15.RELEASE.jar:3.0.15.RELEASE]
        at org.thymeleaf.spring5.view.ThymeleafView.render(ThymeleafView.java:190) ~[thymeleaf-spring5-3.0.15.RELEASE.jar:3.0.15.RELEASE]
```

**Comment by rmathis** [comment]
> Latest commit causes the following:
> 
> ```
> 2022-11-23 09:34:34.754 DEBUG 28299 --- [nio-8080-exec-8] o.s.web.servlet.DispatcherServlet        : Failed to complete request: org.thymeleaf.exceptions.TemplateInputException: Error resolving template [index], template might not exist or might not be accessible by any of the configured Template Resolvers
> 2022-11-23 09:34:34.756 ERROR 28299 --- [nio-8080-exec-8] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.thymeleaf.exceptions.TemplateInputException: Error resolving template [index], template might not exist or might not be accessible by any of the configured Template Resolvers] with root cause
> 
> org.thymeleaf.exceptions.TemplateInputException: Error resolving template [index], template might not exist or might not be accessible by any of the configured Template Resolvers
>         at org.thymeleaf.engine.TemplateManager.resolveTemplate(TemplateManager.java:869) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
>         at org.thymeleaf.engine.TemplateManager.parseAndProcess(TemplateManager.java:607) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
>         at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1098) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
>         at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1072) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
>         at org.thymeleaf.spring5.view.ThymeleafView.renderFragment(ThymeleafView.java:366) ~[thymeleaf-spring5-3.0.15.RELEASE.jar:3.0.15.RELEASE]
>         at org.thymeleaf.spring5.view.ThymeleafView.render(ThymeleafView.java:190) ~[thymeleaf-spring5-3.0.15.RELEASE.jar:3.0.15.RELEASE]
> ```

@mmoayyed I am unable to reproduce this, how did you run the app to get this? I need more details in order to reproduce and debug.

**Comment by mmoayyed** [comment]
I started from scratch. I cloned the repo, switched to this branch, and ran `bootRun` from the root. Then went to `http://localhost:8080/ui`.

**Comment by rmathis** [comment]
> I started from scratch. I cloned the repo, switched to this branch, and ran `bootRun` from the root. Then went to `http://localhost:8080/ui`.

@mmoayyed This explains the issue... my pull request is coming from my fork of the repo, not the branch on this repo. Do you want me to push my changes to this repo's `feature/initializr-ui` and open a new pull request from there?

**Comment by mmoayyed** [comment]
When I checkout this pull request, I am checking out your branch. [Any pull request can be checked out as a branch](https://devopscube.com/checkout-git-pull-request/). These are your latest commits, yes?

image

Regardless, the result is the same if I actually do clone your repo and branch.

**Comment by rmathis** [comment]
> When I checkout this pull request, I am checking out your branch. [Any pull request can be checked out as a branch](https://devopscube.com/checkout-git-pull-request/). These are your latest commits, yes?
> 
> image
> 
> Regardless, the result is the same if I actually do clone your repo and branch.

Ok. I am unable to reproduce this, I pulled it down separately as well and ran this: 
```./gradlew clean build && ./gradlew :app:bootRun```

I am able to view the running app this way. Is there any error during the build with the ui resources? Is there an index.html file in your `app/build/resources/main/templates` folder?

**Comment by mmoayyed** [comment]
Gotcha, thanks. I ended up cleaning up everything by force, and force-cleaned the Gradle cache. Now I am back up! Super weird. Thank you!

Follow up tasks:

- Please default the version to the most recent stable release. For example, now it would be 6.6.2 for the CAS overlay type. 
- Sort the versions array in descending mode where there is more than 1 entry. Larger numbers should stay closer to the top.
- In the footer, add a space between "Powered By" and "Apereo CAS". 

That should be all. After we merge this, next thing would to work on some sort of "preview" functionality. 

Outstanding work. Very nice!

**Comment by mmoayyed** [comment]
There is one issue that needs to be solved still, and that is, 

```bash
./gradlew clean build
java -jar app/build/libs/app.jar
```

...presents the same error as before. This is something to do with the gradle task, and should be fixed. (This is now the app runs in production) I can help with this part.

**Comment by mmoayyed** [comment]
```
rootProject.tasks["wrapper"].finalizedBy([copyWrapper, copyWrapperExecutables]) 
processResources.dependsOn([copyUITemplates, copyUIStatic])
build.dependsOn([copyWrapper, copyWrapperExecutables])
```

If you put these into your `app/build.gradle` towards the end, both modes should work now.

**Comment by rmathis** [comment]
> Gotcha, thanks. I ended up cleaning up everything by force, and force-cleaned the Gradle cache. Now I am back up! Super weird. Thank you!
> 
> Follow up tasks:
> 
>     * Please default the version to the most recent stable release. For example, now it would be 6.6.2 for the CAS overlay type.
> 
>     * Sort the versions array in descending mode where there is more than 1 entry. Larger numbers should stay closer to the top.
> 
>     * In the footer, add a space between "Powered By" and "Apereo CAS".
> 
> 
> That should be all. After we merge this, next thing would to work on some sort of "preview" functionality.
> 
> Outstanding work. Very nice!

This work is completed

**Pull request merged**